
<!-- saved from url=(0094)https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html -->
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">


<script type="text/javascript" src="./ARIADNE V1.2 README_files/analytics.js"></script>
<script type="text/javascript">archive_analytics.values.server_name="wwwb-app15.us.archive.org";archive_analytics.values.server_ms=328;</script>
<link type="text/css" rel="stylesheet" href="./ARIADNE V1.2 README_files/banner-styles.css">


   
   <meta name="GENERATOR" content="Mozilla/4.05C-SGI [en] (X11; I; IRIX 6.5 IP22) [Netscape]">
   <meta name="Author" content="Richard Mott">
   <meta name="Description" content="ARIADNE V1.2 README">

<title>ARIADNE V1.2 README</title>
</head><body text="#000000" bgcolor="#FFFFFF" link="#000099" vlink="#551A8B" alink="#FF0000"><div id="StayFocusd-infobar" style="display: none; top: 0px;">
    <img src="chrome-extension://laankejkbhbdhmipfmgcngdelahlfoji/common/img/eye_19x19_red.png">
    <span id="StayFocusd-infobar-msg"></span>
    <span id="StayFocusd-infobar-links">
        <a id="StayFocusd-infobar-never-show">hide forever</a>&nbsp;&nbsp;|&nbsp;&nbsp;
        <a id="StayFocusd-infobar-hide">hide once</a>
    </span>
</div><div id="wm-ipp" lang="en" style="display: none;">

<div style="position:fixed;left:0;top:0;width:100%!important">
<div id="wm-ipp-inside">
   <table style="width:100%;"><tbody><tr>
   <td id="wm-logo">
       <a href="https://web.archive.org/web/" title="Wayback Machine home page"><img src="./ARIADNE V1.2 README_files/wayback-toolbar-logo.png" alt="Wayback Machine" width="110" height="39" border="0"></a>
   </td>
   <td class="c">
       <table style="margin:0 auto;"><tbody><tr>
       <td class="u" colspan="2">
       <form target="_top" method="get" action="https://web.archive.org/web/form-submit.jsp" name="wmtb" id="wmtb"><input type="text" name="url" id="wmtbURL" value="http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html" style="width:400px;" onfocus="this.focus();this.select();"><input type="hidden" name="type" value="replay"><input type="hidden" name="date" value="20160411141752"><input type="submit" value="Go"><span id="wm_tb_options" style="display:block;"></span></form>
       </td>
       <td class="n" rowspan="2">
           <table><tbody>
           <!-- NEXT/PREV MONTH NAV AND MONTH INDICATOR -->
           <tr class="m">
           	<td class="b" nowrap="nowrap">
		
		    <a href="https://web.archive.org/web/20090803181345/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html" title="3 Aug 2009">AUG</a>
		
		</td>
		<td class="c" id="displayMonthEl" title="You are here: 14:17:52 Apr 11, 2016">APR</td>
		<td class="f" nowrap="nowrap">
		
		    <a href="https://web.archive.org/web/20160711112755/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html" title="11 Jul 2016"><strong>JUL</strong></a>
		
                </td>
	    </tr>
           <!-- NEXT/PREV CAPTURE NAV AND DAY OF MONTH INDICATOR -->
           <tr class="d">
               <td class="b" nowrap="nowrap">
               
                   <a href="https://web.archive.org/web/20090803181345/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html" title="18:13:45 Aug 3, 2009"><img src="./ARIADNE V1.2 README_files/wm_tb_prv_on.png" alt="Previous capture" width="14" height="16" border="0"></a>
               
               </td>
               <td class="c" id="displayDayEl" style="width:34px;font-size:24px;" title="You are here: 14:17:52 Apr 11, 2016">11</td>
	       <td class="f" nowrap="nowrap">
               
		   <a href="https://web.archive.org/web/20160711112755/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html" title="11:27:55 Jul 11, 2016"><img src="./ARIADNE V1.2 README_files/wm_tb_nxt_on.png" alt="Next capture" width="14" height="16" border="0"></a>
	       
	       </td>
           </tr>
           <!-- NEXT/PREV YEAR NAV AND YEAR INDICATOR -->
           <tr class="y">
	       <td class="b" nowrap="nowrap">
               
                   <a href="https://web.archive.org/web/20090803181345/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html" title="3 Aug 2009"><strong>2009</strong></a>
               
               </td>
               <td class="c" id="displayYearEl" title="You are here: 14:17:52 Apr 11, 2016">2016</td>
	       <td class="f" nowrap="nowrap">
               
                   2017
               
	       </td>
           </tr>
           </tbody></table>
       </td>
       </tr>
       <tr>
       <td class="s">
           <a class="t" href="https://web.archive.org/web/20160411141752*/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html" title="See a list of every capture for this URL">16 captures</a>
           <div class="r" title="Timespan for captures of this URL">6 Mar 01 - 11 Jul 16</div>
       </td>
       <td class="k">
       <a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html" id="wm-graph-anchor">
       <div id="wm-ipp-sparkline" title="Explore captures for this URL">
	 <img id="sparklineImgId" alt="sparklines" onmouseover="__wm.st(1)" onmouseout="__wm.st(0)" onmousemove="__wm.mv(event,this)" width="525" height="27" border="0" src="./ARIADNE V1.2 README_files/graph.jsp">
       <div class="yt" style="display: none; width: 25px; height: 27px;"></div><div class="mt" style="display: none; width: 2px; height: 27px;"></div></div>
       </a>
       </td>
       </tr></tbody></table>
   </td>
   <td class="r">
       <a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#close" onclick="__wm.h();return false;" style="background-image:url(/static/images/toolbar/wm_tb_close.png);top:5px;" title="Close the toolbar">Close</a>
       <a href="http://faq.web.archive.org/" style="background-image:url(/static/images/toolbar/wm_tb_help.png);bottom:5px;" title="Get some help using the Wayback Machine">Help</a>
   </td>
   </tr></tbody></table>
</div>
</div>
</div>


<!-- BEGIN WAYBACK TOOLBAR INSERT -->
<script type="text/javascript" src="./ARIADNE V1.2 README_files/disclaim-element.js"></script>
<script type="text/javascript" src="./ARIADNE V1.2 README_files/graph-calc.js"></script>
<script type="text/javascript">//<![CDATA[
var __wm = (function(imgWidth,imgHeight,yearImgWidth,monthImgWidth){
var wbPrefix = "/web/";
var wbCurrentUrl = "http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html";

var firstYear = 1996;
var displayDay = "11";
var displayMonth = "Apr";
var displayYear = "2016";
var prettyMonths = ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"];
var $D=document,$=function(n){return document.getElementById(n)};
var trackerVal,curYear = -1,curMonth = -1;
var yearTracker,monthTracker;
function showTrackers(val) {
  if (val===trackerVal) return;
  var $ipp=$("wm-ipp");
  var $y=$("displayYearEl"),$m=$("displayMonthEl"),$d=$("displayDayEl");
  if (val) {
    $ipp.className="hi";
  } else {
    $ipp.className="";
    $y.innerHTML=displayYear;$m.innerHTML=displayMonth;$d.innerHTML=displayDay;
  }
  yearTracker.style.display=val?"inline":"none";
  monthTracker.style.display=val?"inline":"none";
  trackerVal = val;
}
function trackMouseMove(event,element) {
  var eventX = getEventX(event);
  var elementX = getElementX(element);
  var xOff = Math.min(Math.max(0, eventX - elementX),imgWidth);
  var monthOff = xOff % yearImgWidth;

  var year = Math.floor(xOff / yearImgWidth);
  var monthOfYear = Math.min(11,Math.floor(monthOff / monthImgWidth));
  // 1 extra border pixel at the left edge of the year:
  var month = (year * 12) + monthOfYear;
  var day = monthOff % 2==1?15:1;
  var dateString = zeroPad(year + firstYear) + zeroPad(monthOfYear+1,2) +
    zeroPad(day,2) + "000000";

  $("displayYearEl").innerHTML=year+firstYear;
  $("displayMonthEl").innerHTML=prettyMonths[monthOfYear];
  // looks too jarring when it changes..
  //$("displayDayEl").innerHTML=zeroPad(day,2);
  var url = wbPrefix + dateString + '/' +  wbCurrentUrl;
  $("wm-graph-anchor").href=url;

  if(curYear != year) {
    var yrOff = year * yearImgWidth;
    yearTracker.style.left = yrOff + "px";
    curYear = year;
  }
  if(curMonth != month) {
    var mtOff = year + (month * monthImgWidth) + 1;
    monthTracker.style.left = mtOff + "px";
    curMonth = month;
  }
}
function hideToolbar() {
  $("wm-ipp").style.display="none";
}
function bootstrap() {
  var $spk=$("wm-ipp-sparkline");
  yearTracker=$D.createElement('div');
  yearTracker.className='yt';
  with(yearTracker.style){
    display='none';width=yearImgWidth+"px";height=imgHeight+"px";
  }
  monthTracker=$D.createElement('div');
  monthTracker.className='mt';
  with(monthTracker.style){
    display='none';width=monthImgWidth+"px";height=imgHeight+"px";
  }
  $spk.appendChild(yearTracker);
  $spk.appendChild(monthTracker);

  var $ipp=$("wm-ipp");
  $ipp&&disclaimElement($ipp);
}
return{st:showTrackers,mv:trackMouseMove,h:hideToolbar,bt:bootstrap};
})(525, 27, 25, 2);//]]>
</script>
<style type="text/css">
body {
  margin-top:0 !important;
  padding-top:0 !important;
  min-width:800px !important;
}
</style>

<script type="text/javascript">__wm.bt();</script>
<!-- END WAYBACK TOOLBAR INSERT -->



<center><img src="./ARIADNE V1.2 README_files/wellcome.gif" alt="Wellcome Logo" height="73" width="76" align="CENTER"><font size="+2">
</font><i><font color="#000000"><font size="+1">The Wellcome Trust Centre
for Human Genetics<img src="./ARIADNE V1.2 README_files/crest.gif" alt="OU Crest" height="63" width="58" align="CENTER"></font></font></i></center>

<center><i><font color="#000000"><font size="+1">University of Oxford</font></font></i>&nbsp;
<hr noshade="" width="100%"></center>


<center>
<h1> ARIADNE V1.2 README</h1>

</center>

<p>
Copyright Richard Mott, 2000, <br>
University of Oxford, <br>
Wellcome Trust Centre for Human Genetics, <br>
Roosevelt Drive OX3 7BN UK
</p><p>
Richard.Mott@well.ox.ac.uk

</p><hr noshade="" width="100%">
<h1>Table of Contents</h1>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#introduction">introduction</a><br>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#ariadne">ariadne</a><br>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#prospero">prospero</a><br>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#formats">file formats</a><br>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#install">downloading and installing</a><br>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#run_ariadne">running ariadne</a><br>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#output">output</a><br>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#stats">statistical model</a><br>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#run_prospero">running prospero</a><br>
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#bugs">known bugs and limitations</a><br>
<hr noshade="" width="100%">

<a name="introduction"><h1>Introduction</h1></a>
<p>

ARIADNE is a package of two programs, <b>ariadne</b> and <b>prospero</b>, that compare protein sequences and profiles
using the Smith-Waterman algorithm, and assesses statistical
significance using a new accurate formula, described in Mott, 2000, <a href="http://align_stats2.pdf/">"Accurate Formula for P-values of gapped local sequence and profile alignments"</a> <em>J. Mol Biol.</em> <b>300:</b>649-659<a href="https://web.archive.org/web/20160411141752/http://www.idealibrary.com/links/doi/10.1006/jmbi.2000.3875"> [link to JMB] </a>.

</p><p>
The package is written in ANSI C.
</p><p>
The sequence/profile comparison algorithms used in this code are fairly
standard, and are probably not the fastest implementations
available. The novel part is the method for determining statistical
significance, which will give thresholds of significance that are
accurate to within 5% 95% of the time. See <a href="http://align_stats2.pdf/">Mott (2000)</a> and Mott and
Tribe (1999) for more details.
</p><p>
You are free to incorporate the method used for assessing statistical
significance into third-party code, provided you cite the above reference.  The routines for assessing
significance are all in gaplib.c
</p><p>



</p><hr noshade="" width="100%">
<a name="ariadne"><h3> ariadne </h3></a>

is designed for database searches, and can operate in several modes:
<p></p><ul>
<li> Sequences vs Sequences:	Compare a library of sequence(s) against another</li>
<li> Sequences vs Self: 		Compare a library of sequence(s) against itself</li>
<li> Sequences vs Profiles:		Compare a library of sequence(s) against a library of profile(s)</li>
<li> Profiles vs Sequences:   	Compare a library of profile(s) against a library of sequence(s)</li>
</ul>
(the last two modes produce the same results but in different orders)
<p>
A "library" contains one or more sequences or profiles (see <a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#formats">formats</a>).
</p><p>
ariadne differs from most other database-search programs in that
<a href="https://web.archive.org/web/20160411141752/http://www.well.ox.ac.uk/rmott/ARIADNE/onepage.html#stats">statistical significance</a> is determined on the fly, and statistically
significant alignments are printed out in the order the comparisons
were made, ie they are not sorted. A Perl script ariadne_sort.pl is
provided to sort the output in best-first order.


</p><hr noshade="" width="100%">
<a name="prospero"><h3> prospero </h3></a>

will compare a sequence to itself, another sequence or a profile, and
print all local alignments with p-values less than some user-defined
threshold. Thus prospero is ideal for the analysis of repeats within a
sequence.
<p>

</p><hr noshade="" width="100%">
<a name="formats"><h3>Input File Formats</h3></a>


Substitution matrices (required for sequence to sequence comparisons)
must be in BLAST format. [ARIADNE will look for matrices in the
directory pointed to by the environment variable BLASTMAT if it is
defined.]
<p>
Sequence libraries must be in FASTA format. e.g
</p><pre>&gt;NP_005359
mglsdgewqlvlnvwgkveadipghgqevlirlfkghpetlekfdkfkhlksedemkase
dlkkhgatvltalggilkkkghheaeikplaqshatkhkipvkylefiseciiqvlqskh
pgdfgadaqgamnkalelfrkdmasnykelgfqg
</pre>

Format for profiles:
<p>
A library of profiles can contain multiple profiles.
Format for each profile is a 3-line header followed by the profile data. Blank lines are ignored.
The header looks like
</p><pre>&gt;name             		name of the profile, eg "globin"
profile-length alphabet-length  eg "141 21"
alphabet-order                  List of space-separated amino-acids in the order they occur in the profile
</pre>
The profile contains a series of lines, one per profile position, with format
<p>
[position] [consensus] [profile-scores in order defined by alphabet-order]
</p><p>
For example, part of a profile looks like this:
</p><pre>&gt;globin
141 21
A R N D C Q E G H I L K M F P S T W Y V X
    1 H   -4  -1   2  -2  -5  -1  -1  -3  10  -5  -5  -2  -4  -3  -4  -2  -3  -4   0  -5 -1
    2 L   -4  -5  -6  -7   0  -5  -6  -6  -5  -1   6  -5   0   0  -6  -5  -4   6  -3  -2 -1
    3 S   -1  -4  -1  -1  -4  -2  -2  -2  -4  -4  -4  -3  -2  -5  -4   5   6  -5  -5  -3 -1
    4 A    4  -3   0   3  -1   0   2  -1  -2  -3  -3  -1  -2  -3   0   0  -2  -6  -3  -3 -1
    5 E    2  -1  -1   3  -3   0   3  -1   0  -5  -3   1  -4  -3  -1  -1  -1  -6  -4  -3 -1
    6 E   -3  -3   0   4  -6   4   5  -4  -3  -3  -5  -2  -4  -6  -4  -2  -1  -6  -5  -3 -1
    7 K    0   3  -2  -4  -2  -1   1  -4  -1   0  -3   5  -2  -1  -4  -2  -3   4  -4  -1 -1
    8 A    3  -1   0   0  -2   2   0  -2   1  -2  -2   1  -3  -2  -4   1   1  -5  -4  -3 -1
    9 L    2  -2   2  -2  -2  -2  -1  -4   0   1   3   0  -1  -2  -5  -2   0  -5  -2   0 -1
   10 V   -2  -6  -6  -6  -4  -5  -5  -6  -6   5   1  -5  -1  -2  -5  -4  -3  -6  -4   6 -1
   11 K   -1   3   1  -2  -3   2  -2  -3   0  -3   0   5   0  -5  -4  -1   1  -5  -4  -2 -1
   12 A    2  -1   0   1  -3   0   0   0   1  -4  -3   1  -2  -4  -4   3   0  -6  -4  -3 -1
   13 L    0  -4  -2  -3   0   0  -2  -4  -1   1   1  -3  -1  -1  -4   3   2  -5  -3   1 -1
   14 W   -3  -6  -6  -5  -2  -5  -4  -4  -5  -2  -4  -6  -2   2  -7  -4  -4  13   0  -2 -1
........... etc etc
</pre>
So eg at position 2 the consensus is L and the score for matching an A at this position is -4

<p>
Profiles and substitution matrices should be such that the smallest
span of score values is 1 Also, for reasons of efficiency it is best
is the total range of score values is quite small, say &lt; 30
</p><p>
Some example profile sets derived from PFAM-A are provided for <a href="http://pfam.profiles.z/">download</a>.
</p><p>

</p><hr noshade="" width="100%">
<a name="install"><h3>Downloading and Installing ARIADNE</h3></a>

A compressed tarred source distribution is available for download <a href="http://ariadne-1.2.tar.z/">here</a>. 
<p>
Alternatively, you can download precompiled binaries for <a href="http://ariadne-1.2-linux.tar.z/">Linux</a>, <a href="http://ariadne-1.2-irix64.tar.z/">SGI IRIX64</a>, <a href="http://ariadne-1.2-sunos.tar.z/">SunOS</a>, <a href="http://ariadne-1.2-osf.tar.z/">OSF</a>.
 
</p><p>
To compile the sources, after untarring the distribution ariadne-1.2.tar.Z type the following
commands
</p><pre>% cd ./SRC-1.2/
% make
</pre>
The binaries will be in the directory ./SRC-1.2/$UNAME
where $UNAME is the machine architecture returned by the uname
command.  Either put this directory on your path or copy the binaries
to a directory in your path.<p>

<b>NB</b> You will need a version of make compatible with gnu make. If you don't have this, then you can compile the programs with the following commands executed in the source directory:
</p><pre>cc -o ./ariadne -O -I./Include cl.c gaplib.c profile_util.c prospero_utils.c searchpath.c cmp.c profile_stat.c ariadne.c readline.c seq_util.c -lm 

cc -o ./prospero -O -I./Include cl.c gaplib.c profile_util.c prospero_utils.c searchpath.c cmp.c profile_stat.c prospero.c readline.c seq_util.c -lm 
</pre>

This will leave the executables in the source directory; they will need to be moved elsewhere by hand.
<p>
</p><hr noshade="" width="100%">
<a name="run_ariadne"><h3>Running ariadne</h3></a>

<pre>usage: ariadne
        -mode           text                 [  ]
        -seq            Readable File        [  ]
        -seq2           Readable File        [  ]
        -profile        Readable File        [  ]
        -matrix         text                 [  ]
        -A              integer              [ 11 ]
        -B              integer              [ 1 ]
        -ethresh        float                [ 0.1 ]
        -dbsize         float                [ 1 ]
        -help           switch               [  ]      
</pre>
<p>
</p><ul>
<li>-mode defines the type the search. It must be one of the following:
<ul>
<li>ss2self		Compare a file of sequences against themselves, omitting duplicated and self-comparisons
</li><li>s2ss		Compare one sequence against a library of sequences
</li><li>ss2ss		Compare a library of sequences against another library
</li><li>s2pp		Compare one sequence against a library of profiles
</li><li>p2ss		Compare one profile against a library of sequences
</li><li>pp2ss		Compare a library of profiles against a library of sequences
</li><li>ss2pp		Compare a library of sequences against a library of profiles
</li></ul><p>
It is always faster to compare profile(s) to sequence(s), because of
the overhead of reading the profiles, so use mode pp2ss in preference to ss2pp.
</p><p>

</p></li><li>-seq  {file}		The first FASTA file of sequences 
</li><li>-seq2 {file}		The second FASTA file of sequences (only required for modes s2ss or ss2ss ) NOTE: -seq specifies the query sequence and -seq2 the database in mode s2ss.
</li><li>-profile {file}		The file of profiles


</li><li>-matrix {matrix}        The name of the matrix

Defines the substitution matrix. Only required for modes ss2self,
s2ss, ss2ss. ariadne uses the BLAST matrix distribution, and looks in
$BLASTMAT/{matrix}

</li><li>-A {gap-open penalty}
</li><li>-B {gap-extend penalty}

The gap penalties. NOTE: a gap of length k residues is scored as
A+B*k.  This convention is different from FASTA's which uses
A+B*(k-1). So e.g. the default FASTA gap penalty of 12, 2 corresponds
to A=10, B=2.

</li><li>-dbsize {integer}

The effective size, in <b>SEQUENCES</b>, of the database. You MUST set
this number, otherwise it defaults to a database size of 1
sequence. Because ariadne prints out matches as it finds them, it does
not know the database size ahead of time. Therefore you must provide
this information explicitly. Note that the database size differs from
the database length (used in BLAST).
<ul>
<li> You can set the dbsize to an artificial value such as 100000 so that
you can compare the results of different searches more easily.  This
corresponds to fixing the size of the protein universe. Because of the
redundancy (or near-redundancy) of many protein sequence databanks, it
is not necessarily correct to set dbsize to the total number of
sequences, because this is overly conservative. The "correct" value is
the number of independent comparisons made in the search.

</li><li> If you are comparing one library, size M, against another, size N,
you may want to set dbsize=M*N.

</li><li> If you are self-comparing a library of size N you may want to set
dbsize=N*(N-1)/2

</li><li> The database size is used to help determine the statistical
significance of matches (see below).
</li></ul>

</li><li>-ethresh {float}

The e-value threshold is used to define the cutoff for printing out
alignments. The default value of ethresh is 0.1. 
Suppose a comparison has a pairwise p-value P, taking
account of the sequence/profile lengths, composition and scoring
scheme. The evalue E is defined as
<p>
E = P*dbsize 
</p><p>
and if E &lt; ethresh the alignment is printed out
</p><p>

<b>WARNING:</b> If you set ethresh to a high value (or set dbsize to a low value) then
the search will take considerably longer. This is because statistical
significance is assessed initially assuming a standard sequence
composition, so that pre-computed parameters are used. If the
estimated e-value is &lt; 10*ethresh then statistical significance is
recomputed using parameters that reflect the
sequence/profile compositions accurately. This second computation is quite slow,
but is usually only triggered in a small proportion of cases, provided
the dbsize and ethresh are set appropriately.
</p><p>
If you want the pairwise p-values printed, set the dbsize to 1 and set
the ethresh to a small value such as 1.0e-8. You will then get all
similarities with p-values &lt; 1.0e-8. This will still be fast.
</p></li></ul>

<hr noshade="" width="100%">
<a name="output"><h3>Output</h3></a>


ariadne prints out statistically significant alignments in the order
that it finds them. Each alignment is preceeded by a one-line summary,
e.g.
<pre>&gt; 0 d1bal__ len 51 from 14 to 49  vs  d2pdd__ len 43 from 5 to 40   score 77  eval 6.954146e-10 identity 36.11% K 1.361261e-01 L 3.478619e-01 H 1.187679e+00 alpha 8.223475e-02

   14 PAIRRLLAEHNLDASAIKGTGVGGRLTREDVEKHLA    49  d1bal__
      |::|:   |  :|   ::|||  ||: :||::  ||
    5 PSVRKYAREKGVDIRLVQGTGKNGRVLKEDIDAFLA    40  d2pdd__

</pre>
The one-line summary says that the comparison between d1bal__ (51
residues) and d2pdd__ (43 residues) scored 77 with an e-value of
6.954e-10 and a percentage identity 36%. The parameters used for
assessing significance, K, L, H, alpha took the values 0.136, 0.347,
1.187, 0.082. [H, alpha are ony printed for information].
<p>
Profile-Sequence alignments look the same as Sequence-Sequence
alignments, with the profile consensus printed.
</p><p>
</p><hr noshade="" width="100%">
<a name="stats"><h3>Statistical Model</h3></a>


ariadne and prospero test the statistical significance of similarity scores using
the following model ( <a href="http://align_stats2.pdf/">download this</a> for more details):
<p>
A score S in a comparison between two sequences or a sequence and a
profile has pairwise p-value P given by
</p><p>
P = 1-exp(-Kmn exp(-L*S)) 
</p><p>
where m,n are the sequence lengths and K, L are parameters depending
on the compositions, scoring scheme and (slightly) on the sequence
lengths. K, L are calculated using the formula described in Mott,
2000, which takes account of sequence composition, substitution
matrix, gap penalty, sequence length:
</p><p>
L =  L_u*(1.013 -2.61*alpha + f(m,n)( -0.76 + 9.34*alpha +1.12/H) )</p><p>
K = K_u*exp( 0.26 -18.92*alpha + f(m,n)(-1.76 + 32.69*alpha + 192.52*alpha^2 + 3.24/H ) )
</p><p>
where: f(m,n) = log(m*n)*(1/m+1/n) L_u, K_u are the parameters for
ungapped alignments H is the entropy of ungapped alignments (eg as
defined Karlin-Altschul PNAS 1990, or Mott and Tribe 1999), and alpha
is a parameter depending on the gap penalty A+B*k:
</p><p>
alpha = 2*s*exp(-L_u(A+B))/(1-exp(-L_u*B))
</p><p>
where 
</p><p>
s = sqrt{ (K_u/H) * [ delta / exp(L_u *delta ) ] }
</p><p>
and delta is the smallest span of score values (usually 1) 
</p><p>

In the example above, K = 1.361261e-01 L = 3.478619e-01, m = 51, n = 43, S = 125, so  
</p><p>
P = 1-exp(-0.1361*51*43*exp(-0.3478*125)) = 5.95e-16. The database
size was set at 100000 = 1.0e6 sequences, so the evalue is 5.95e-10.
</p><p>

</p><hr noshade="" width="100%">
<a name="run_prospero"><h3>Running prospero</h3></a>

Most of the command-line switches are the same as for ariadne:
<pre>usage: prospero
	-seq1           Readable File        [  ]
	-seq2           Readable File        [  ]
	-pro            Readable File        [  ]
	-matrix         text                 [ blosum62 ]
	-A              integer              [ 11 ]
	-B              integer              [ 1 ]
	-top            integer              [ 5 ]
	-pseudocount    integer              [ 100 ]
	-ethresh        float                [ 0 ]
	-help           switch               [  ]
</pre>
<ul>
<li> -seq1, -seq2 define the paths to FASTA-format files containing the two sequences to be compared.
</li><li> -pro defines the path to the profile.

If you specify only -seq1 then the sequence is compared to itself, and the end-to-end alignment is suppressed).
If you specify only -seq1 and -pro then the sequence is compared against the profile

</li><li>-matrix defines the substitution matrix, which is needed for sequence-sequence comparisons only (default is blosum62)

</li><li>-A, -B defines the gap penalty. (see ariadne) (default 11+1k)

</li><li>-top defines how many alignments to print (default is 5). -top 1 will print just the optimal local alignment

</li><li>-ethresh defines the p-value threshold; if this is set then only similarities with smaller p-values are printed. Overides -top

</li><li>-pseudocount. You should not need to alter this parameter. It controls the length of the virtual sequence whose composition is added to the observed compositions to prevent over-fitting, which can result in a loss of sensitivity.
</li></ul>
<p>
Output from prospero is very similar to ariadne's. The E-values reported are in fact P-values, as the database size is set to 1. The alignments are sorted by evalue. Example:

</p><pre>
pangloss [84]% ./SRC-1.2/Linux/prospero -seq1 gi130316.pep -ethresh 0.001
***** PROSPERO V1.0  Sun May 14 13:25:04 2000 *****

Copyright Richard Mott, 2000, Wellcome Trust Centre for Human Genetics, University of Oxford

using gap penalty 11+1k
using matrix blosum62
printing all alignments with eval &lt; 0.001000
using sequence1 gi|130316|sp|P00747|PLMN_HUMAN
using self-comparison

&gt; 1 gi|130316|sp|P00747|PLMN_HUMAN len 810 from 102 to 455  vs  gi|130316|sp|P00747|PLMN_HUMAN len 810 from 184 to 561   score 994  eval 7.000317e-93 identity 49.58% K 1.978546e-02 L 2.230037e-01 H 1.439763e+00 alpha 1.068759e-01

  102 ECKTGNGKNYRGTMSKTKNGITCQKWSSTSPHRPRFSPATHPSEGLEENYCRNPDNDPQG   161  gi|130316|sp|P00747|PLMN_HUMAN
      ||   :|:|| | :||| :|: || | | |||   : |:  |:: |::||||||| : : 
  184 ECMHCSGENYDGKISKTMSGLECQAWDSQSPHAHGYIPSKFPNKNLKKNYCRNPDRELR-   242  gi|130316|sp|P00747|PLMN_HUMAN

  162 PWCYTTDPEKRYDYCDILECEE---------ECMHCSGENYDGKISKTMSGLECQAWDSQ   212  gi|130316|sp|P00747|PLMN_HUMAN
      |||:|||| ||:: |||  |           :|:  :|||| | :: |:||  || | :|
  243 PWCFTTDPNKRWELCDIPRCTTPPPSSGPTYQCLKGTGENYRGNVAVTVSGHTCQHWSAQ   302  gi|130316|sp|P00747|PLMN_HUMAN

  213 SPHAHGYIPSKFPNKNLKKNYCRNPDRELRPWCFTTDPNKRWELCDIPRC----------   262  gi|130316|sp|P00747|PLMN_HUMAN
      :|| |   |  || ||| :||||||| :  ||| ||:   ||| | || |          
  303 TPHTHNRTPENFPCKNLDENYCRNPDGKRAPWCHTTNSQVRWEYCKIPSCDSSPVSTEQL   362  gi|130316|sp|P00747|PLMN_HUMAN

  263 -TTPPPSSGPTYQ-CLKGTGENYRGNVAVTVSGHTCQHWSAQTPHTHNRTPENFPCKNLD   320  gi|130316|sp|P00747|PLMN_HUMAN
        | ||   |  | |  | |::|||  : | :|  || ||: ||| | :||||:|   | 
  363 APTAPPELTPVVQDCYHGDGQSYRGTSSTTTTGKKCQSWSSMTPHRHQKTPENYPNAGLT   422  gi|130316|sp|P00747|PLMN_HUMAN

  321 ENYCRNPDGKRAPWCHTTNSQVRWEYCKIPSCDSSPVSTEQLAPTA--PPELTPVVQDCY   378  gi|130316|sp|P00747|PLMN_HUMAN
       |||||||  : ||| ||:  |||||| :  |  :  |     |    |   ||  :|| 
  423 MNYCRNPDADKGPWCFTTDPSVRWEYCNLKKCSGTEASVVAPPPVVLLPDVETPSEEDCM   482  gi|130316|sp|P00747|PLMN_HUMAN

  379 HGDGQSYRGTSSTTTTGKKCQSWSSMTPHRHQ-KTPENYPNAGLTMNYCRNPDAD-KGPW   436  gi|130316|sp|P00747|PLMN_HUMAN
       |:|: |||  :|| ||  || |::  ||||   |||  | |||  ||||||| |  |||
  483 FGNGKGYRGKRATTVTGTPCQDWAAQEPHRHSIFTPETNPRAGLEKNYCRNPDGDVGGPW   542  gi|130316|sp|P00747|PLMN_HUMAN

  437 CFTTDPSVRWEYCNLKKCS   455  gi|130316|sp|P00747|PLMN_HUMAN
      |:||:|   ::||:: :|:
  543 CYTTNPRKLYDYCDVPQCA   561  gi|130316|sp|P00747|PLMN_HUMAN

&gt; 2 gi|130316|sp|P00747|PLMN_HUMAN len 810 from 102 to 352  vs  gi|130316|sp|P00747|PLMN_HUMAN len 810 from 274 to 560   score 670  eval 1.676058e-61 identity 48.00% K 1.978546e-02 L 2.230037e-01 H 1.439763e+00 alpha 1.068759e-01

  102 ECKTGNGKNYRGTMSKTKNGITCQKWSSTSPHRPRFSPATHPSEGLEENYCRNPDNDPQG   161  gi|130316|sp|P00747|PLMN_HUMAN
      :|  | |:|||| :: | :| ||| ||: :||    :|   | : |:||||||||   : 
  274 QCLKGTGENYRGNVAVTVSGHTCQHWSAQTPHTHNRTPENFPCKNLDENYCRNPDG-KRA   332  gi|130316|sp|P00747|PLMN_HUMAN

  162 PWCYTTDPEKRYDYCDILECE---------------------EECMHCSGENYDGKISKT   200  gi|130316|sp|P00747|PLMN_HUMAN
      |||:||: : |::|| |  |:                     ::| |  |::| |  | |
  333 PWCHTTNSQVRWEYCKIPSCDSSPVSTEQLAPTAPPELTPVVQDCYHGDGQSYRGTSSTT   392  gi|130316|sp|P00747|PLMN_HUMAN

  201 MSGLECQAWDSQSPHAHGYIPSKFPNKNLKKNYCRNPDRELRPWCFTTDPNKRWELCDIP   260  gi|130316|sp|P00747|PLMN_HUMAN
       :| :||:| | :|| |   |  :||  |  ||||||| :  ||||||||: ||| |:: 
  393 TTGKKCQSWSSMTPHRHQKTPENYPNAGLTMNYCRNPDADKGPWCFTTDPSVRWEYCNLK   452  gi|130316|sp|P00747|PLMN_HUMAN

  261 RCT-------TPPP-------SSGPTYQCLKGTGENYRGNVAVTVSGHTCQHWSAQTPHT   306  gi|130316|sp|P00747|PLMN_HUMAN
      :|:        |||        :     |: | |: |||  | ||:|  || |:|| || 
  453 KCSGTEASVVAPPPVVLLPDVETPSEEDCMFGNGKGYRGKRATTVTGTPCQDWAAQEPHR   512  gi|130316|sp|P00747|PLMN_HUMAN

  307 HN-RTPENFPCKNLDENYCRNPDGK-RAPWCHTTNSQVRWEYCKIPSC   352  gi|130316|sp|P00747|PLMN_HUMAN
      |:  |||  |   |::||||||||    |||:||| :  ::|| :| |
  513 HSIFTPETNPRAGLEKNYCRNPDGDVGGPWCYTTNPRKLYDYCDVPQC   560  gi|130316|sp|P00747|PLMN_HUMAN

&gt; 3 gi|130316|sp|P00747|PLMN_HUMAN len 810 from 100 to 288  vs  gi|130316|sp|P00747|PLMN_HUMAN len 810 from 374 to 587   score 467  eval 7.668240e-42 identity 47.34% K 1.978546e-02 L 2.230037e-01 H 1.439763e+00 alpha 1.068759e-01

  100 LSECKTGNGKNYRGTMSKTKNGITCQKWSSTSPHRPRFSPATHPSEGLEENYCRNPDNDP   159  gi|130316|sp|P00747|PLMN_HUMAN
      : :|  |:|::|||| | |  |  || ||| :||| : :|  :|: ||  ||||||| | 
  374 VQDCYHGDGQSYRGTSSTTTTGKKCQSWSSMTPHRHQKTPENYPNAGLTMNYCRNPDAD-   432  gi|130316|sp|P00747|PLMN_HUMAN

  160 QGPWCYTTDPEKRYDYCDILEC-----------------------EEECMHCSGENYDGK   196  gi|130316|sp|P00747|PLMN_HUMAN
      :||||:||||  |::||:: :|                       ||:||  :|: | ||
  433 KGPWCFTTDPSVRWEYCNLKKCSGTEASVVAPPPVVLLPDVETPSEEDCMFGNGKGYRGK   492  gi|130316|sp|P00747|PLMN_HUMAN

  197 ISKTMSGLECQAWDSQSPHAHG-YIPSKFPNKNLKKNYCRNPDREL-RPWCFTTDPNKRW   254  gi|130316|sp|P00747|PLMN_HUMAN
       : |::|  || | :| || |  : |   |   |:|||||||| ::  |||:||:| | :
  493 RATTVTGTPCQDWAAQEPHRHSIFTPETNPRAGLEKNYCRNPDGDVGGPWCYTTNPRKLY   552  gi|130316|sp|P00747|PLMN_HUMAN

  255 ELCDIPRCTTPPPSSG-PTYQCLKGTGENYRGNVA   288  gi|130316|sp|P00747|PLMN_HUMAN
      : ||:|:|  |    | |  :  |  |    | ||
  553 DYCDVPQCAAPSFDCGKPQVEPKKCPGRVVGGCVA   587  gi|130316|sp|P00747|PLMN_HUMAN

&gt; 4 gi|130316|sp|P00747|PLMN_HUMAN len 810 from 102 to 196  vs  gi|130316|sp|P00747|PLMN_HUMAN len 810 from 480 to 575   score 284  eval 4.056174e-24 identity 50.53% K 1.978546e-02 L 2.230037e-01 H 1.439763e+00 alpha 1.068759e-01

  102 ECKTGNGKNYRGTMSKTKNGITCQKWSSTSPHRPR-FSPATHPSEGLEENYCRNPDNDPQ   160  gi|130316|sp|P00747|PLMN_HUMAN
      :|  |||| |||  : |  |  || |::  |||   |:| |:|  |||:||||||| |  
  480 DCMFGNGKGYRGKRATTVTGTPCQDWAAQEPHRHSIFTPETNPRAGLEKNYCRNPDGDVG   539  gi|130316|sp|P00747|PLMN_HUMAN

  161 GPWCYTTDPEKRYDYCDILECEEECMHCSGENYDGK   196  gi|130316|sp|P00747|PLMN_HUMAN
      |||||||:| | |||||: :|      |     : |
  540 GPWCYTTNPRKLYDYCDVPQCAAPSFDCGKPQVEPK   575  gi|130316|sp|P00747|PLMN_HUMAN

pangloss [85]% 
</pre>

<hr noshade="" width="100%">

<h3> Known Bugs and limitations </h3>

Currently the smallest span of score values in the substitution matrix or profile must be 1. (Will be fixed shortly)
<p>
Ariadne will only find the optimal alignment between each pair of sequences - ie it does a single-pass Smith-Waterman comparison. Prospero will find all suboptimal alignments. Ariadne will be modified in the next release to find all significant subalignments.
</p><p>
The formula for statistical significance has been thoroughly tested on
standard substition matrices (ie the PAM and BLOSUM series). If you
use it on other matrices/profiles and get unexpected results please contact me.
</p><p>
</p><hr noshade="" width="100%">

<address>
Please send Questions, Comments, and Bug Reports to <a href="mailto:richard.mott@well.ox.ac.uk">Richard Mott </a></address>
<hr noshade="" width="100%">








<!--
     FILE ARCHIVED ON 14:17:52 Apr 11, 2016 AND RETRIEVED FROM THE
     INTERNET ARCHIVE ON 14:28:59 Dec 4, 2016.
     JAVASCRIPT APPENDED BY WAYBACK MACHINE, COPYRIGHT INTERNET ARCHIVE.

     ALL OTHER CONTENT MAY ALSO BE PROTECTED BY COPYRIGHT (17 U.S.C.
     SECTION 108(a)(3)).
-->
</body></html>